Add new DeviceReader constructors

bruno-f-cruz opened the pull request:
Closes #25

bruno-f-cruz committed:
Add new constructor methods

bruno-f-cruz committed:
Add new constructor methods

bruno-f-cruz committed:
Add new constructor methods

bruno-f-cruz committed:
Deprecate function

bruno-f-cruz committed:
Deprecate function

bruno-f-cruz committed:
Deprecate function

bruno-f-cruz committed:
Document methods

bruno-f-cruz committed:
Document methods

bruno-f-cruz committed:
Document methods

bruno-f-cruz committed:
Linting

bruno-f-cruz committed:
Linting

bruno-f-cruz committed:
Linting

glopesdev commented on harp/reader.py:
```diff
@@ -75,6 +80,245 @@ def __dir__(self) -> Iterable[str]:
+    @staticmethod
+    def from_url(
+        url: str,
+        base_path: Optional[PathLike] = None,
```
`base_path` is indeed getting clunky, I'm slightly uncomfortable of having different behaviors depending on which function you use here, so maybe just getting rid of it would be best.

bruno-f-cruz replied:
Not sure I follow. The use of `base_path` is consistent across methods. If provided it will try use it as a root directory where all files are expected to be found. If not provided it will assume either the current working directory (e.g. `from_url`) or the root of the provided `yml` (e.g. `from_file`). Are these two defaults what are raising flags for you? Happy to go with a different one if you want. We can also just throw an error at the level of the `DeviceReader` class if no path was provided and the user try to use the no-input overload of the register load function.

glopesdev reviewed (changes requested):
@bruno-f-cruz thanks, looks good overall. I have a few questions / comments:

- Would be nice if [bf816e2](https://github.com/harp-tech/harp-python/pull/26/commits/bf816e288fc68d00b12228aef03e779deecca666) would be in its own PR so that it gets listed in the release notes. Is it too hard to extract? Doesn't look like the rest of the PR is depending on it.
- What linter are you using? Can you run black on the current repo state? When I run it on my end I'm getting a different result. Maybe we could agree on what tools we use and then add them to `pyproject.toml` in a separate PR.
- I'm having second thoughts about the use of `base_path`. I don't like that it is used in different ways depending on which function gets called, so maybe it would be best to get rid of it entirely?

glopesdev commented on harp/reader.py:
```diff
@@ -75,6 +80,245 @@ def __dir__(self) -> Iterable[str]:
+        return DeviceReader(device, reg_readers)
+
+    @staticmethod
+    def from_str(
```
Still looking for a better name:
- `from_yaml`: may be misleading since the others also ultimately use YAML
- `from_schema`: maybe a little bit better

bruno-f-cruz replied:
`from_schema` makes me nervous as people might think you should pass the actuall schema file (i.e. `.yml`). Would `from_string` be better (`ToString` is also a thing after all haha)?

bruno-f-cruz commented on harp/model.py:
```diff
@@ -6,16 +6,10 @@
-)
+from pydantic import (BaseModel, BeforeValidator, ConfigDict, Field, RootModel,
+                      field_serializer)
+from typing_extensions import Annotated
```
Fixed it by running black.

Regarding the order I am tempted to leave it as is, as it is the suggested sorting order given by isort (the new linter I included in the project.toml). Let me know if this makes sense. If not we can remove isort as a depedendency and go with what we had

bruno-f-cruz commented on harp/reader.py:
```diff
@@ -1,17 +1,23 @@
+from harp import __version__
 from harp.io import MessageType, read
+from harp.model import BitMask, GroupMask, Model, PayloadMember, Register
 from harp.schema import read_schema
```
You are correct, all other changes are from running isort.

bruno-f-cruz commented:
@glopesdev lets review #28 and #27 first and then we will handle this one.

bruno-f-cruz committed:
Merge branch 'feat-device-reader-constructors' of https://github.com/harp-tech/harp-python into feat-device-reader-constructors

bruno-f-cruz committed:
Fix rebasing

bruno-f-cruz committed:
Add deprecated decorator

bruno-f-cruz committed:
Add deprecated decorator

bruno-f-cruz committed:
Fix rebasing

bruno-f-cruz committed:
Favor library's deprecated decorator

bruno-f-cruz committed:
Linting

bruno-f-cruz committed:
Linting

bruno-f-cruz committed:
Add requests as dependency

bruno-f-cruz committed:
Add requests as dependency

bruno-f-cruz committed:
Move decorator to separate private helper module to prevent circular dependencies

bruno-f-cruz committed:
Move decorator to separate private helper module to prevent circular dependencies

bruno-f-cruz committed:
Ignore warnings emitted by deprecated decorator

bruno-f-cruz committed:
Ignore warnings emitted by deprecated decorator

bruno-f-cruz committed:
Merge branch 'main' into feat-device-reader-constructors

bruno-f-cruz committed:
Favor stdlib decorator

bruno-f-cruz committed:
Favor stdlib decorator

bruno-f-cruz committed:
Refactor to #40

bruno-f-cruz committed:
Refactor to #40

bruno-f-cruz committed:
Linting

bruno-f-cruz commented:
@glopesdev I am taking a second look at this and I think we need to mainly decide on a few points:

1. Method's names (currently):
-`from_model`(takes a parsed instance of the device schema, ie. `Model` class)
-`from_str`(takes the literal yml file)
-`from_url`(takes an URL to a device yml)
-`from_file`(takes the file path to a local device yml)
-`from_dataset`(takes the path to a "on-standard" container, replaces the current entry point)

2. Are any of the methods redundant? (looking at from_file vs from_str for instance)
3. The fate of base_path. I still think it serves a purpose in the API of the function, but maybe you have a cleaner way in mind.

bruno-f-cruz committed:
Make `data` input to `_compose_parser` optional

This allows `g` aka `_create_register_reader` to take care of the path inference logic

bruno-f-cruz committed:
Add unittest for registers with payload spec

bruno-f-cruz committed:
Suppress `pyright` error on internal method import

glopesdev committed:
Merge pull request #46 from harp-tech/fix-parser

Make `data` input to `_compose_parser` optional

glopesdev committed:
Replace license classifier with SPDX expression

glopesdev committed:
Merge pull request #47 from harp-tech/license-spdx

Replace license classifier with SPDX expression

bruno-f-cruz committed:
Merge branch 'feat-device-reader-constructors' of https://github.com/harp-tech/harp-python into feat-device-reader-constructors

bruno-f-cruz committed:
Linting

bruno-f-cruz committed:
Import deprecated from typing_extensions

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/harp-tech/harp-python?pullRequest=26) <br/>All committers have signed the CLA.

bruno-f-cruz committed:
Ensure device_name is appended to base_path

bruno-f-cruz committed:
Add unittest

bruno-f-cruz committed:
Add string type hinting to paths

bruno-f-cruz committed:
Refactor unions to legacy syntax

bruno-f-cruz committed:
Ensure request errors are raised from reader

bruno-f-cruz committed:
Shorten deprecation warning
